File: Problem115_the-maze.java
// BFS
// Time Complexity: O(mn * max(m, n)) = O(mn * constant) = O(mn)
// Space Complexity: O(mn)
import java.util.*;
class Solution {
    public boolean hasPath(int[][] maze, int[] start, int[] destination) {
        // null
        if(maze == null) return false;
        int m = maze.length, n = maze[0].length;
        Queue<int[]> q = new LinkedList<>();
        q.add(start);
        maze[start[0]][start[1]] = 2;
        int[][] dirs = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}}; // right, left, bottom, top
        while(!q.isEmpty()) {
            int[] curr = q.poll();
            for(int[] dir: dirs) {
                int row = curr[0];
                int col = curr[1];
                // hit the wall
                while(row >= 0 && col >= 0 && row < m && col < n && maze[row][col] != 1) {
                    row = row + dir[0];
                    col = col + dir[1];
                }
                // come back one step
                row = row - dir[0];
                col = col - dir[1];
                // wherever it stops
                if(maze[row][col] == 0) {
                    if(row == destination[0] && col == destination[1])
                        return true;
                    maze[row][col] = 2;
                    q.add(new int[]{row, col});
                }
            }
        }
        return false;
    }
}

File: Problem-115 The Maze.py
# 490. The Maze
# https://leetcode.com/problems/the-maze/

# Logic: We apply a variuation of BFS. Rather then just adding 
# the neighbouring nodes, we keep on exploring all the possible 
# nodes in the direction of the neighbours till we hit out of 
# bounds or a wall. The place we stop acts as a child node to 
# the parent and gets added to the queue. Before adding it we 
# check if it is the destination, if it is, we return true else 
# continue. After the q is empty, we return false.


# Time Complexity: O(n*m(m+n))
# Space Complexity: O(n*m)

from collections import deque

class Solution:
    def hasPath(self, maze, start, dest):
        n = len(maze)
        m = len(maze[0])

        q = deque()
        
        q.append(start)
        maze[start[0]][start[1]] = 2
        
        dirs = ((0,-1),(0,1),(-1,0), (1,0))
        
        while q:
            row, col = q.popleft()
            
            for _dir in dirs:
                newRow = row
                newCol = col
                
                while 0 <= newRow < n and 0 <= newCol < m and maze[newRow][newCol] != 1:
                    newRow += _dir[0]
                    newCol += _dir[1]
                
                newRow -= _dir[0]
                newCol -= _dir[1]

                if (newRow, newCol) == (dest[0], dest[1]):
                    return True
                
                if maze[newRow][newCol] != 2:
                    maze[newRow][newCol] = 2
                    q.append((newRow, newCol))

        return False